### PR TITLE
WIP: dns/bind: full zone management

### PR DESCRIPTION
**Pull request by mimugmail**
Hi, first draft for full zone management. There are two new tabs, one for domains:

image

And then a second tab for resource records with all records for domains like A = 1.2.3.4 etc.
In records I have a ModelRelationField linking this record to a domain. 

image

I'm a bit confused about templating. I need a file for every zone, this is easy as you see below. Also the zone values like ttl's are find. But I have no idea how to get the linked records? 
Do I need to flip the first for loop with records and get all values from UUID? Perhaps it's too easy and I cant find it :)

As usual please dont review as tons of stuff is missing.

### CODE REVIEW

**Comment by AdSchellevis** [comment]
@mimugmail if you would like to discuss some ideas about grid input, just ping me on irc. It should be rather easy to create some master/detail views which doesn't require to select the domain in every grid again.  We do similar things with the same pattern for other applications as well (for example choosing the customer in the first tab, extra details bound to the first in the second)

**Comment by mimugmail** [comment]
Some nested grid would be nice. I just build this WIP to have a PoC and base for discussion. :)

**Comment by fabianfrz** [comment]
@mimugmail just look at the nginx sources - for example SNI hostname map. It has a custom grid in grid.

**Comment by mimugmail** [comment]
I already did, but I dont understand the JS foo :)

**Comment by AdSchellevis** [comment]
@mimugmail that's not the pattern I'm suggesting, let's talk about this on irc. Maybe I can help you out with a simple example

File: dns/bind/src/opnsense/mvc/app/controllers/OPNsense/Bind/Api/RecordController.php [hunk 1]
```diff
@@ -38,12 +38,32 @@ class RecordController extends ApiMutableModelControllerBase
...
+        $domain = $this->request->get('domain');
+        $result = $this->getBase('record', 'records.record', $uuid);
+        if ($uuid == null && !empty($result['record']['domain'])) {
+            // set domain selection
+            foreach ($result['record']['domain'] as $key => &$value) {
+                if ($key == $domain) {
+                    $value['selected'] = 1;
+                } else {
+                    $value['selected'] = 0;
+                }
```
Comment (fabianfrz) [inline dns/bind/src/opnsense/mvc/app/controllers/OPNsense/Bind/Api/RecordController.php]:
```php
$value['selected'] = $key == $domain ? 1 : 0;
```

File: dns/bind/src/opnsense/mvc/app/views/OPNsense/Bind/general.volt [hunk 2]
```diff
@@ -177,24 +172,53 @@ $( document ).ready(function() {
...
-            'get':'/api/bind/record/getRecord/',
-            'set':'/api/bind/record/setRecord/',
-            'add':'/api/bind/record/addRecord/',
-            'del':'/api/bind/record/delRecord/',
-            'toggle':'/api/bind/record/toggleRecord/'
+    }).on("selected.rs.jquery.bootgrid", function(e, rows) {
+        $("#grid-records").bootgrid('reload');
+    }).on("deselected.rs.jquery.bootgrid", function(e, rows) {
+        $("#grid-records").bootgrid('reload');
+    }).on("loaded.rs.jquery.bootgrid", function (e) {
```
Comment (fabianfrz) [inline dns/bind/src/opnsense/mvc/app/views/OPNsense/Bind/general.volt]: if you don't need the parameters, you can omit them

File: dns/bind/src/opnsense/mvc/app/views/OPNsense/Bind/general.volt [hunk 3]
```diff
@@ -110,6 +172,53 @@ $( document ).ready(function() {
...
+        'add':'/api/bind/record/addRecord/',
+        'del':'/api/bind/record/delRecord/',
+        'toggle':'/api/bind/record/toggleRecord/',
+        options:{
+            useRequestHandlerOnGet: true,
+            requestHandler: function() {
+                let request = {'domain': 'not_found'};
+                let ids = $("#grid-domains").bootgrid("getSelectedRows");
+                if (ids.length > 0) {
+                    request['domain'] = ids[0];
```
Comment (fabianfrz) [inline dns/bind/src/opnsense/mvc/app/views/OPNsense/Bind/general.volt]:
```suggestion
                    request.domain = ids[0];
```

File: dns/bind/src/opnsense/mvc/app/views/OPNsense/Bind/general.volt [hunk 4]
```diff
@@ -1,7 +1,7 @@
 {#
 
-OPNsense® is Copyright © 2014 – 2018 by Deciso B.V.
-This file is Copyright © 2018 by Michael Muenz <m.muenz@gmail.com>
+OPNsense® is Copyright © 2014 – 2019 by Deciso B.V.
+This file is Copyright © 2018 - 2019 by Michael Muenz <m.muenz@gmail.com>
```
Comment (fabianfrz) [inline dns/bind/src/opnsense/mvc/app/views/OPNsense/Bind/general.volt]: maybe use the same dash here as in the previous line

File: dns/bind/src/opnsense/mvc/app/models/OPNsense/Bind/Record.xml [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+                    <default></default>
+                    <Required>N</Required>
+                </name>
+                <type type="OptionField">
+                    <default>A</default>
+                    <Required>Y</Required>
+                    <OptionValues>
+                        <A>A</A>
+                        <CNAME>CNAME</CNAME>
+                        <MX>MX</MX>
```
Comment (fabianfrz) [inline dns/bind/src/opnsense/mvc/app/models/OPNsense/Bind/Record.xml]: you will not get far without SRV-, TXT and CAA records. I would also recommend to support NS records.

Comment (mimugmail) [reply dns/bind/src/opnsense/mvc/app/models/OPNsense/Bind/Record.xml]: As I said ... please don't review, there are tons of features missing. Only opened to get some help regarding templating :)

**Comment by mimugmail** [comment]
Closing this draft PR until I'm finished, thank you @AdSchellevis :)

**Pull request closed without merging**